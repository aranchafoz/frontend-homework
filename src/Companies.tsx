import { connect, useDispatch } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import { ReduxState, Company } from './types'
import { getCompanies } from './selectors'
import { setSelectedCompanyId, toggleDropdownMenuVisibility } from './actions'

import CompanyLink from './CompanyLink'

type ReduxProps = {
  companies: Array<Company>,
}

type DispatchProps = {
  setSelectedCompanyId: (id: number) => void,
  toggleDropdownMenuVisibility: () => void,
}

export const Companies = ({ companies }: ReduxProps & DispatchProps) => {

  const dispatch = useDispatch();

  function handleSelectCompany(id: number) {
    dispatch(setSelectedCompanyId(id));
    dispatch(toggleDropdownMenuVisibility());
  }

  return (
    <>
      <div>Your companies</div>

      {companies.map((company) => (
        <CompanyLink
          key={company.id}
          {...company}
          onClick={(id) => handleSelectCompany(id)}
        />
      ))}
    </>
  )
}

export default connect(
  createStructuredSelector<ReduxState, ReduxProps>({
    companies: getCompanies,
  }),
  {
    setSelectedCompanyId,
    toggleDropdownMenuVisibility,
  }
)(Companies)
