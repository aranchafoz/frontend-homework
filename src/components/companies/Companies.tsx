import { connect, useDispatch } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import { ReduxState, Company } from '../../types'
import { getCompanies } from '../../redux/selectors'
import { setSelectedCompanyId, toggleDropdownMenuVisibility } from '../../redux/actions'

import CompanyLink from '../company-link/CompanyLink'

type ReduxProps = {
  companies: Array<Company>,
}

type DispatchProps = {
  setSelectedCompanyId: (id: number) => void,
  toggleDropdownMenuVisibility: () => void,
}

export const Companies = ({ companies }: ReduxProps & DispatchProps) => {

  const dispatch = useDispatch();

  function handleSelectCompany(id: number) {
    dispatch(setSelectedCompanyId(id));
    dispatch(toggleDropdownMenuVisibility());
  }

  return (
    <>
      <div>Your companies</div>

      {companies.map((company) => (
        <CompanyLink
          key={company.id}
          {...company}
          onClick={(id) => handleSelectCompany(id)}
          {...{[`data-test-company-link-${company.id}`]: true}}
        />
      ))}
    </>
  )
}

export default connect(
  createStructuredSelector<ReduxState, ReduxProps>({
    companies: getCompanies,
  }),
  {
    setSelectedCompanyId,
    toggleDropdownMenuVisibility,
  }
)(Companies)
