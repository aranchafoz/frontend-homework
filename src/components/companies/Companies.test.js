import { shallow } from 'enzyme'
import { Companies } from './Companies'
import * as redux from "react-redux";

describe('<Companies />', () => {
  const useDispatchMock = jest.spyOn(redux, 'useDispatch')
  const companies = [
    { id: 1, name: 'Dummy company' },
    { id: 2, name: 'Smarty company' },
  ]

  const render = () => shallow(
    <Companies companies={companies} />
  )

  beforeEach(() => {
    useDispatchMock.mockClear()
  })

  it('renders list of company links', () => {
    expect(render()).toMatchSnapshot()
  })

  it('renders list of companies does not call useDispatch', function(){
    const dummyDispatch = jest.fn()
    useDispatchMock.mockReturnValue(dummyDispatch)

    //action
    render()

    //assert
    expect(dummyDispatch).not.toHaveBeenCalled()
  })

  it('renders list of companies call useDispatch', function(){
    const dummyDispatch = jest.fn()
    useDispatchMock.mockReturnValue(dummyDispatch)

    //action
    const component = render()
    component.find('[data-test-company-link-1]').simulate('click')

    //assert
    expect(dummyDispatch).toHaveBeenCalled()
  })
})
