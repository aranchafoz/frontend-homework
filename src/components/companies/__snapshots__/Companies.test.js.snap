// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Companies /> renders list of company links 1`] = `
<Fragment>
  <div>
    Your companies
  </div>
  <CompanyLink
    data-test-company-link-1={true}
    id={1}
    key="1"
    name="Dummy company"
    onClick={[Function]}
  />
  <CompanyLink
    data-test-company-link-2={true}
    id={2}
    key="2"
    name="Smarty company"
    onClick={[Function]}
  />
</Fragment>
`;
